python/ray/util/queue.py

```diff
@@ -87,22 +131,76 @@ def get(self, block=True, timeout=None):
             else:
                 return ray.get(self.actor.get.remote(timeout))
 
+    async def get_async(self, block=True, timeout=None):
+        """Gets an item from the queue.
+
+        There is no guarantee of order if multiple consumers get from the
+        same empty queue.
+
+        Returns:
+            The next item in the queue.
+        Raises:
+            Empty: if the queue is empty and blocking is False.
+            Empty: if the queue is empty, blocking is True, and it timed out.
+            ValueError: if timeout is negative.
+        """
+        if not block:
+            try:
+                return await self.actor.get_nowait.remote()
+            except asyncio.QueueEmpty:
+                raise Empty
+        else:
+            if timeout is not None and timeout < 0:
+                raise ValueError("'timeout' must be a non-negative number")
+            else:
+                return await self.actor.get.remote(timeout)
+
     def put_nowait(self, item):
         """Equivalent to put(item, block=False).
 
         Raises:
-            Full if the queue is full.
+            Full: if the queue is full.
         """
         return self.put(item, block=False)
 
+    def put_nowait_batch(self, items):
+        """Takes in a list of items and puts them into the queue in order.
+
+        Raises:
+            Full: if the items will not fit in the queue
+        """
+        if not isinstance(items, list):
+            raise TypeError("Argument 'items' must be a list")
+        if self.maxsize > 0 and len(items) + self.qsize() > self.maxsize:
+            raise Full("Cannot add " + str(len(items)) +
+                       " items to queue of size " + str(self.qsize()) +
+                       " and maxsize " + str(self.maxsize))
+        ray.get(self.actor.put_nowait_batch.remote(items))
```

Comment:
This is not thread-safe.  Move this logic into the nested Ray Actor method.  Ray Actor methods are guaranteed to be executed start-to-finish with no switching.